### PR TITLE
Fixed adding Trees/Corn/Wine via dynamic script

### PR DESCRIPTION
**Pull request by vqw**
ScriptTryObjectSet fixed, checking for allowable objects as well (we do
not want falling trees etc)

### CODE REVIEW

File: src/terrain/KM_Terrain.pas [hunk 1]
```diff
@@ -548,6 +548,15 @@ function TKMTerrain.ScriptTryObjectSet(X, Y: Integer; aObject: Byte): Boolean;
       end;
   end;
 
+  function AllowableObject : Boolean;
```
Comment (Kromster80) [inline src/terrain/KM_Terrain.pas:551]: Please add a comment upfront, what is a "AllowableObject"

File: src/terrain/KM_Terrain.pas [hunk 2]
```diff
@@ -555,20 +564,58 @@ function TKMTerrain.ScriptTryObjectSet(X, Y: Integer; aObject: Byte): Boolean;
...
   DiagonalChanged := MapElem[Land[Y,X].Obj].DiagonalBlocked xor MapElem[aObject].DiagonalBlocked;
 
   //Apply change
   Land[Y, X].Obj := aObject;
-  UpdatePassability(KMRect(X, Y, X, Y)); //When using KMRect map bounds are checked by UpdatePassability
-  UpdateWalkConnect([wcWalk, wcRoad, wcWork], KMRectGrowTopLeft(KMRect(X, Y, X, Y)), DiagonalChanged);
-  Result := True;
+  case aObject of
+
+    55..58:
```
Comment (Kromster80) [inline src/terrain/KM_Terrain.pas]: Please add comments on to each case statement, what are these magic numbers

**Comment by vqw** [comment]
Comments added

File: src/terrain/KM_Terrain.pas [hunk 3]
```diff
@@ -555,20 +564,59 @@ function TKMTerrain.ScriptTryObjectSet(X, Y: Integer; aObject: Byte): Boolean;
...
+      else
+        Result := False;
+    88..124, // Trees - 125 is mushroom
+    126..172:
+    begin
+      if ObjectIsChopableTree(KMPoint(X,Y), caAge1) then Land[Y,X].TreeAge := 1;
+      if ObjectIsChopableTree(KMPoint(X,Y), caAge2) then Land[Y,X].TreeAge := TREE_AGE_1;
+      if ObjectIsChopableTree(KMPoint(X,Y), caAge3) then Land[Y,X].TreeAge := TREE_AGE_2;
+      if ObjectIsChopableTree(KMPoint(X,Y), caAgeFull) then Land[Y,X].TreeAge := TREE_AGE_FULL;
+      UpdatePassability(KMRect(X, Y, X, Y)); //When using KMRect map bounds are checked by UpdatePassability
```
Comment (Kromster80) [inline src/terrain/KM_Terrain.pas]: Why do we update passability and walk connect only for trees and other objects?

**Comment by vqw** [comment]
UpdatePassability and UpdateWalkConnect is used in SetField so that we only use it in trees and other objects

File: src/terrain/KM_Terrain.pas [hunk 4]
```diff
@@ -555,20 +564,60 @@ function TKMTerrain.ScriptTryObjectSet(X, Y: Integer; aObject: Byte): Boolean;
...
+      else
+        Result := False;
+    59..63: // Corn in different stages
+      if CanAddField(X, Y, ft_Corn) then
+      begin
+        SetField(KMPoint(X, Y), -1, ft_Corn);
+        Result := True;
+      end
+      else
+        Result := False;
```
Comment (Kromster80) [inline src/terrain/KM_Terrain.pas:601]: When we return False, does this imply that the whole operation failed? And if it has failed, then maybe it should not have the "Land[Y, X].Obj := aObject;" set.

**Comment by vqw** [comment]
Fixed this, along with placing Wine/Corn on coal deposits which was wrong.
TileGoodForField does not check this - did it manually in this corner case.

**Comment by Kromster80** [comment]
When we return False from ScriptTryObjectSet, does this imply that the whole operation failed? And if it has failed, then maybe it should not have the "Land[Y, X].Obj := aObject;" set.

**Comment by vqw** [comment]
"Land[Y, X].Obj := aObject;" moved inside each case statement (please check the src) and when cannot place field it returns False and means that whole operation failed.

**Comment by Kromster80** [comment]
For the future, please create new pull requests separately.

**Pull request merged**